tests/test_IssuanceController.py

```
@@ -0,0 +1,432 @@
+
+        # Transfer the amount to the receiver
+        txr = self.issuanceController.exchangeEtherForNomins(exchanger, ethToSend)
+
+        # Ensure the result of the transfer is correct.
+        endingFundsWalletEthBalance = get_eth_balance(self.fundsWallet)
+        self.assertEqual(self.nomin.balanceOf(self.issuanceControllerContract.address), nominsBalance)
+        self.assertEqual(self.nomin.balanceOf(exchanger), 0)
+        self.assertEqual(self.nomin.feePool(), 0)
+        self.assertEqual(startingFundsWalletEthBalance, endingFundsWalletEthBalance)
+
+    def test_exchangeFailsNotEnoughNomins(self):
+        amount = 10 * UNIT
+        nominsBalance = (int(amount / 2) * self.usdToEthPrice) // UNIT
+        base = self.nomin.amountReceived(nominsBalance)
+        feesToPayInNomins = nominsBalance - base
+        exchanger = self.participantAddresses[0]
+        startingFundsWalletEthBalance = get_eth_balance(self.fundsWallet)
+
+        # Set up the contract so it contains some nomins for folks to convert Ether for
+        self.nomin.giveNomins(MASTER, self.issuanceControllerContract.address, nominsBalance)
+        self.assertEqual(self.nomin.balanceOf(self.issuanceControllerContract.address), nominsBalance)
+
+        # Ensure the transfer fails due to there not being enough nomins in the contract
+        self.assertReverts(self.issuanceController.exchangeEtherForNomins, exchanger, amount)
+
+        # Ensure the result of the transfer is correct.
+        endingFundsWalletEthBalance = get_eth_balance(self.fundsWallet)
+        self.assertEqual(self.nomin.balanceOf(self.issuanceControllerContract.address), nominsBalance)
+        self.assertEqual(self.nomin.balanceOf(exchanger), 0)
+        self.assertEqual(self.nomin.feePool(), 0)
+        self.assertEqual(startingFundsWalletEthBalance, endingFundsWalletEthBalance)
+
+    # Withdraw havvens tests
+
+    def test_withdrawHavvens(self):
```

Maybe worth testing that the transaction reverts if the user attempts to withdraw too many havvens (rather than just failing silently). Also partial withdrawals.

And the same thing for `test_withdrawNomins`.